### ISSUE TITLE
Improve performance of pulse.schedule

### ISSUE THREAD

**Issue by itoko** [Qiskit/qiskit-terra#2817]
### What is the expected enhancement?
Construction of `Schedule` using `append` should be much faster.
Currently, it takes 34s (273s) for constructing a schedule with 2000 (4000) instructions. Those are  measured by `create_random_appended_schedule(100, 20)` (200, 20) using the following code: 

```
def create_random_appended_schedule(n_instrs_by_channel, n_channels, pulse_size=10):
    channels = [pulse.DriveChannel(i) for i in range(n_channels)]
    instrs = []
    for i in range(n_instrs_by_channel):
        command = pulse.SamplePulse(np.random.random(pulse_size))
        for chan in channels:
            instrs.append(command(chan))

    start_time = time.time()
    schedule = pulse.Schedule()
    for instr in instrs:
        schedule += instr
    end_time = time.time()
    print(end_time-start_time)
```

For drastic speed up, we may need a lot of refactoring, but I think we can improve the performance somewhat by a small modification.

**Comment by itoko** [comment]
I'm tackling this issue.

**Comment by mtreinish** [comment]
Do you think it makes sense to add some pulse benchmarks for cases like this to the asv benchmark suite? https://github.com/Qiskit/qiskit/tree/master/test/benchmarks If so we can track the performance of these operations over time at https://qiskit.github.io/qiskit/

**Comment by taalexander** [comment]
Yes, I think it makes sense to add something like this as a benchmark. We will likely be refactoring the module soon, so it would be great to see how performance tracks.

**Comment by mtreinish** [comment]
@taalexander @lcapelluto can this be closed since #2995 has merged?

**Comment by taalexander** [comment]
Yes, Closing via #2995.
